runtime/common/src/auctions.rs

```diff
@@ -0,0 +1,1229 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Auctioning system to determine the set of Parachains in operation. This includes logic for the
+//! auctioning mechanism and for reserving balance as part of the "payment". Unreserving the balance
+//! happens elsewhere.
+
+use sp_std::{prelude::*, mem::swap, convert::TryInto};
+use sp_runtime::traits::{CheckedSub, Zero, One, Saturating};
+use parity_scale_codec::Decode;
+use frame_support::{
+	decl_module, decl_storage, decl_event, decl_error, ensure, dispatch::DispatchResult,
+	traits::{Currency, ReservableCurrency, Get, Randomness, EnsureOrigin},
+	weights::{DispatchClass, Weight},
+};
+use primitives::v1::Id as ParaId;
+use frame_system::ensure_signed;
+use crate::slot_range::{SlotRange, SLOT_RANGE_COUNT};
+use crate::traits::{Leaser, LeaseError, Auctioneer};
+
+type CurrencyOf<T> = <<T as Config>::Leaser as Leaser>::Currency;
+type BalanceOf<T> = <<<T as Config>::Leaser as Leaser>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub trait WeightInfo {
+	fn new_auction() -> Weight;
+	fn bid() -> Weight;
+	fn on_initialize() -> Weight;
+}
+
+pub struct TestWeightInfo;
+impl WeightInfo for TestWeightInfo {
+	fn new_auction() -> Weight { 0 }
+	fn bid() -> Weight { 0 }
+	fn on_initialize() -> Weight { 0 }
+}
+
+/// The module's configuration trait.
+pub trait Config: frame_system::Config {
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+
+	/// The number of blocks over which a single period lasts.
+	type Leaser: Leaser<AccountId=Self::AccountId, LeasePeriod=Self::BlockNumber>;
+
+	/// The number of blocks over which an auction may be retroactively ended.
+	type EndingPeriod: Get<Self::BlockNumber>;
```

Comment:
A comment here on configuration would be helpful. When using BABE randomness we need at least 1 full session between the time the bids are collected and the time we end the auction, otherwise the auction end is predictable. 

So if bids are collected in session S, we can use the current epoch randomness (`Babe::random`) in session S+2. This would be configured by having this value be at least `epoch_len_in_slots` blocks, which will take at least that many slots.

Reply:
Might even need to be S+3 as the randomness for S+2 is almost or fully determined at the end of session S. Will let @burdges elaborate.

Alternatively we can change the `Babe::random` implementation to use `NextRandomness` but that has its own set of footguns.

Reply:
https://github.com/paritytech/substrate/pull/8180 will also help with configuration.

Reply:
https://github.com/paritytech/substrate/pull/8180